frontend/lib/resolution/Resolver.cpp

```diff
@@ -3617,6 +3651,60 @@ void Resolver::handleCallExpr(const uast::Call* call) {
   }
 }
 
+// TODO: Handle 'always allow ref' for iterators.
+bool Resolver::enter(const Zip* zip) {
+  auto p = parsing::parentAst(context, zip);
+  const bool requiresParallel = p && p->isForall();
+  const bool prefersParallel = requiresParallel || (p && p->isBracketLoop());
+
+  CHPL_ASSERT(p);
+
+  if (!p || !p->isIndexableLoop()) {
+    context->error(zip, "zip expressions may only appear as a loop's iterand");
+  }
+
+  std::vector<QualifiedType> eltTypes;
+  QualifiedType leaderType;
+
+  for (int i = 0; i < zip->numActuals(); i++) {
+    const bool isFirstActual = (0 == i);
+    auto actual = zip->actual(i);
+
+    if (actual->isZip()) {
+      context->error(actual, "nested 'zip' expressions are not supported");
+      auto& re = byPostorder.byAst(actual);
+      eltTypes.push_back(re.type());
+      continue;
+    }
+
+    // Try to resolve the parallel leader using the first actual.
+    if (isFirstActual && prefersParallel) {
+      const bool emitErrors = requiresParallel;
+      leaderType = resolveParallelLeaderIterType(*this, actual, actual,
```

Comment:
I'm going to change the structure of the underlying implementation so that the resolved iterator can be gotten through an optional out formal. Then you can see inspect it to see the details you're describing.